contrib/babelfishpg_tsql/src/pl_exec-2.c

```diff
@@ -4057,3 +4064,317 @@ exec_stmt_fulltextindex(PLtsql_execstate *estate, PLtsql_stmt_fulltextindex *stm
+
+	/* check if the login has necessary permissions for CREATE/DROP */
+	perform_permission_check(partition_function_name, stmt->is_create, true);
+
+	if (stmt->is_create) 
+	{
+		PLtsql_type		*typ = stmt->datatype;
+		List 			*arg = stmt->args;
+		bool 			isnull;
+		Oid			valtype;
+		int32			valtypmod;
+		Datum			tsql_type_datum;
+		char			*tsql_typename = NULL;
+		Datum			*values = NULL;
+		ArrayType		*arr_value = NULL;
+		Oid			sql_variant_oid;
+		tsql_compare_context 	cxt;
+		Oid			basetype_oid;
+		Oid			opclass_oid;
+		Oid			opfamily_oid;
+		Oid			cmpfunction_oid;
+		int 			nargs = list_length(arg);
+		int			dbid = get_cur_db_id();
+		MemoryContext		cur_ctxt = CurrentMemoryContext;
+		LOCAL_FCINFO(fcinfo, 1);
+
+		/* check if given name is exceeding the allowed limit */
+		if (strlen(partition_function_name) > 128)
+		{
+			ereport(ERROR, 
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					errmsg("The identifier that starts with '%.128s' is too long. Maximum length is 128.", partition_function_name)));
+		}
+
+		/* check if there is existing partition function with the given name in the current database */
+		if (partition_function_exists(partition_function_name, dbid))
+		{
+			ereport(ERROR, 
+				(errcode(ERRCODE_DUPLICATE_FUNCTION),
+					errmsg("There is already an object named '%s' in the database.", partition_function_name)));
+		}
+
+		/* get the tsql typename from the input type */
+		InitFunctionCallInfoData(*fcinfo, NULL, 0, InvalidOid, NULL, NULL);
+		fcinfo->args[0].value = ObjectIdGetDatum(typ->typoid);
+		fcinfo->args[0].isnull = false;
+		tsql_type_datum = (*common_utility_plugin_ptr->translate_pg_type_to_tsql) (fcinfo);
+
+		if (tsql_type_datum)
+		{
+			tsql_typename = text_to_cstring(DatumGetTextPP(tsql_type_datum));
+		}
+
+		/*
+		 * check if datatype is supported or not, if tsql_typename is NULL
+		 * then it implies that typename corresponds to UDT
+		 */
+		if (!tsql_typename || is_tsql_text_ntext_or_image_datatype(typ->typoid) ||
+			(*common_utility_plugin_ptr->is_tsql_geometry_datatype) (typ->typoid) ||
+			(*common_utility_plugin_ptr->is_tsql_geography_datatype) (typ->typoid) ||
+			(*common_utility_plugin_ptr->is_tsql_rowversion_or_timestamp_datatype) (typ->typoid) ||
+			typ->typoid == XMLOID) /* we don't have XML type specific to TSQL */
+		{
+			ereport(ERROR, 
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					errmsg("The type '%s' is not valid for this operation.", typ->typname)));
+		}
+
+		/* check if the given number of boundaries are exceeding allowed limit */
+		if (nargs >= 15000)
+		{
+			ereport(ERROR, 
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					errmsg("CREATE/ALTER partition function failed as only a maximum of 15000 partitions can be created.")));
+		}
+
+		values = palloc0(nargs * sizeof(Datum));
+
+		for (int i = 0; i < nargs; i++)
+		{
+			Datum val;
+
+			/* evaluate the value from the expr */
+			val = exec_eval_expr(estate, list_nth(arg, i), &isnull, &valtype, &valtypmod);
```

Comment:
Do we have enough test cases to cover these scenarios?

Reply:
Lets add more test cases where local variables are getting values from volatile functions